### PR TITLE
PrimeString Solution

### PR DESCRIPTION
**Pull request by arjun9218**
Prime String solution also includes JUnit test

### CODE REVIEW

File: Solutions-1/PrimeString.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	private Logger logger;
+	
+	/**
+	 * Instantiates a new PrimeString
+	 */
+	public PrimeString() {
+		logger = Logger.getLogger(PrimeString.class.getSimpleName());
+	}
+	
+	/**
```
Comment (bsmaheshkumar) [inline Solutions-1/PrimeString.java:24]: which code formate is applied?

**Review by bsmaheshkumar — COMMENTED** [review]
Thanks @arjun9218  for solving this problem.
As part of Review request, could you please include the screen shots of pass test cases and test code coverage?

File: Solutions-1/PrimeString.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	 * Checks if is prime.
+	 *
+	 * @param Integer the number
+	 * @return true, if is prime
+	 */
+	
+	protected boolean isPrime(Integer number) {
```
Comment (bsmaheshkumar) [inline Solutions-1/PrimeString.java:53]: which this method is protected, why not public of private?

File: Solutions-1/PrimeString.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	
+	/**
+	 * Checks if is Prime String.
+	 *
+	 * @param String the input string
+	 * @return true, if is Prime String
+	 */
+	
+	protected String isPrimeString(String input) {
```
Comment (bsmaheshkumar) [inline Solutions-1/PrimeString.java:103]: Only this method can be public rest can be private.

File: Solutions-1/PrimeStringTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	
+	PrimeString instance = new PrimeString();
+	
+	/**
+	 * Before testing the isPrimeString method, we need to make sure the supporting methods are working as expected.
+	 * preRequisites method tests isPrime, buildMapFromString and isPrimeStringLogic methods.
+	 */
+	
+	@Before
+	public void preRequisites() {
```
Comment (bsmaheshkumar) [inline Solutions-1/PrimeStringTest.java:29]: Why do we need this test cases?

File: Solutions-1/PrimeStringTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		boolean logicExpected = true;
+		assertEquals( logicExpected, logicResult);
+	}
+
+	/**
+	 * testing method tests the isPrimeString method.
+	 */
+	
+    @Test
+    public void testing() {
```
Comment (bsmaheshkumar) [inline Solutions-1/PrimeStringTest.java:50]: what about Negative test case?
What is the unit test coverage?

**Review by bsmaheshkumar — CHANGES_REQUESTED** [review]
Need to add more unit test cases, both positive and negative.
Please address the review comments.
Can you convert the same to maven project? And solve two more problem? Before we move on to Angular?

File: Solutions-1/PrimeString.java (hunk 3 above)
Comment (arjun9218) [reply Solutions-1/PrimeString.java:103]: I had made that method as protected because I wanted to test all the methods, I read your review comments and I'll make all the methods private and the isPrimeString method as public.

File: Solutions-1/PrimeString.java (hunk 2 above)
Comment (arjun9218) [reply Solutions-1/PrimeString.java:53]: Yeah sure I'll convert this to maven project and solve two more problems.

File: Solutions-1/PrimeStringTest.java (hunk 4 above)
Comment (arjun9218) [reply Solutions-1/PrimeStringTest.java:29]: I thought of testing all the methods so I had put this. I'll make changes as per your comments.

File: Solutions-1/PrimeString.java (hunk 3 above)
Comment (bsmaheshkumar) [reply Solutions-1/PrimeString.java:103]: Ok.
You can also unit test private method's view public method.
No need to change the access of the method only for unit testing.

**Comment by bsmaheshkumar** [comment]
@neerajkujha @coderhindu 
Adding Neeraj and Narasimhan to review the code.

File: solutuins-1/pom.xml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>solutions-1</groupId>
+  <artifactId>solutuins-1</artifactId>
```
Comment (bsmaheshkumar) [inline solutuins-1/pom.xml:6]: typo solutions

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	 *
+	 * @param int size of the army, Integer[] number of weapons each soldier is carrying
+	 * @return String ready or not ready for battle based on Kattapa's luck logic
+	 */
+
+	public String isReadyOrNot(int size, Integer[] soldierWeapons) {
+		try {
+			int evenWeaponsCount = 0;
+			int oddWeaponsCount = 0;
+			for(int i=0;i<size;i++) {
```
Comment (bsmaheshkumar) [inline solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:33]: Can you apply code format.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/PrimeString.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	
+	/**
+	 * Checks if the number of distinct alphabets and the number of occurrences of each alphabets is prime.
+	 *
+	 * @param Map<Character, Integer> the map that is built from the input string
+	 * @return true, if is Prime String
+	 */
+	
+	private boolean isPrimeStringLogic(Map<Character, Integer> map) {
```
Comment (bsmaheshkumar) [inline solutuins-1/src/main/java/solutions_1/solutuins_1/PrimeString.java:31]: do we need logic in method name, isPrimeString should be enough?

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	/**
+	 * Checks if the even weapon solders are greater than odd weapon soldiers.
+	 *
+	 * @param int size of the army, Integer[] number of weapons each soldier is carrying
+	 * @return String ready or not ready for battle based on Kattapa's luck logic
+	 */
+
+	public String isReadyOrNot(int size, Integer[] soldierWeapons) {
```
Comment (bsmaheshkumar) [inline solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: Can this method be renamed to isReadyForBattle
On need to Use is and or together in function name.

File: solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@RunWith(Parameterized.class)
+public class KattappaTest {
+
+	/**
+	 * Instantiating KattappaTest object
+	 */
+
+	KattapaReadyForBattle instance = new KattapaReadyForBattle();
```
Comment (bsmaheshkumar) [inline solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:26]: can the instance be rename to kattapaBattleTest?

File: solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		Object[][] data = new Object[][] { { 4,new Integer[]{11, 12, 13, 14},"NOT READY" }, { 3,new Integer[]{2, 3, 4},"READY FOR BATTLE" } };
+		return Arrays.asList(data);
+	}
+
+	/**
+	 * testing method tests the isReadyOrNot method.
+	 */
+
+	@Test
+	public void testing() {
```
Comment (bsmaheshkumar) [inline solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:55]: The method name code be like kattapaReadyForBattle_Positive_Test

The verification can be simple like below
```
@Test
public void kattapaReadyForBattle_Positive_Test  () {
   assertEquals("NOT READY", kattapaBattleTest.isReadyForBattle(4, new Integer[]{11, 12, 13, 14}));
   assertEquals("READY FOR BATTLE", kattapaBattleTest.isReadyForBattle(4, new Integer[]{2, 3, 4));
}
```

Comment (bsmaheshkumar) [reply solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:55]: Similar to Positive Add Negative test cases

```
@Test
public void kattapaReadyForBattle_Negative_Test  () {
   assertEquals("Number of Soliders and weapon not matching ", kattapaBattleTest.isReadyForBattle(2, new Integer[]{11, 12, 13, 14}));
   assertEquals("Invalid input", kattapaBattleTest.isReadyForBattle(4, null);
}
```

Please include the junit result image.
And also include test code coverage image.

File: solutuins-1/src/test/java/solutions_1/solutuins_1/NonReducibleTest.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * The Class PrimeStringTest.
+ */
+
+
+@RunWith(Parameterized.class)
+public class NonReducibleTest {
```
Comment (bsmaheshkumar) [inline solutuins-1/src/test/java/solutions_1/solutuins_1/NonReducibleTest.java:20]: Similar to KattapaReadyForBattle test class, please update the tests other classes.

File: solutuins-1/src/test/java/solutions_1/solutuins_1/ShoppingTest.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+	/**
+	 * testing method tests the getRoyPayment method.
+	 */
+
+	@Test
+	public void testing() { 
+		int result = instance.getRoyPayment( shoppingCart );
+		//String expected = "Yes";
+		System.out.println("Expected : " + expected + " | Actual : " + result);
```
Comment (bsmaheshkumar) [inline solutuins-1/src/test/java/solutions_1/solutuins_1/ShoppingTest.java:56]: do we need sys out?

**Review by bsmaheshkumar — CHANGES_REQUESTED** [review]
Few changes in unit test cases required.

File: solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java (hunk 11 above)
Comment (arjun9218) [reply solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:55]: I had included a screenshot. It's a .bmp file.

File: solutuins-1/src/test/java/solutions_1/solutuins_1/NonReducibleTest.java (hunk 12 above)
Comment (arjun9218) [reply solutuins-1/src/test/java/solutions_1/solutuins_1/NonReducibleTest.java:20]: Sure, will do and commit the changes

File: solutuins-1/src/test/java/solutions_1/solutuins_1/ShoppingTest.java (hunk 13 above)
Comment (arjun9218) [reply solutuins-1/src/test/java/solutions_1/solutuins_1/ShoppingTest.java:56]: I was trying to make the Test runs to look little detailed. I had attached snapshots to base directory, I've again added screenshots in a separate folder.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/PrimeString.java (hunk 8 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/PrimeString.java:31]: There is one more method isPrimeString, isPrimeStringLogic is just a supporting method.

File: Solutions-1/PrimeString.java (hunk 1 above)
Comment (arjun9218) [reply Solutions-1/PrimeString.java:24]: I had manually entered the comments, I had not applied code format.

File: Solutions-1/PrimeString.java (hunk 3 above)
Comment (arjun9218) [reply Solutions-1/PrimeString.java:103]: Ok got it.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 7 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:33]: Sure I'll apply code format and commit the latest code

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 9 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: Yeah. I'll rename it

**Comment by arjun9218** [comment]
@simha111 adding Narasimha for code review

File: Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+	public String isPrimeString(String input) {
+		try {
+			Map<Character, Integer> map = buildMapFromString(input);
+			boolean primeString = isPrimeStringLogic(map);
+			if (primeString) {
+				return "Yes";
+			} else {
+				return "No";
+			}
+		} catch (Exception e) {
```
Comment (neerajkujha) [inline Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java:115]: what kind of exception do you expect to be thrown?

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 9 above)
Comment (neerajkujha) [inline solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: why an array of Integer not int?

File: solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+			logger.log(Level.SEVERE, " NonReducible::getChopCount()");
+		}
+		return chopCount;
+	}
+}
+
+/**
+ * The Class ChopCount.
+ */
+class ChopCount {
```
Comment (neerajkujha) [inline solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: why we need a class here? Can we achieve this without this class?

File: Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java (hunk 14 above)
Comment (arjun9218) [reply Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java:115]: I was thinking if the input was null, then NullPointerException may get thrown so I had surrounded the code with try catch.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 9 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: No particular reason here, can use int as well. In PrimeString problem I had used Maps which didn't accept the primitive datatype int so I was using Integer there. In that flow I had used Integer here.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java (hunk 15 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: In the problem statement, the expected output was an object with reduced string and number of operation so I had created such class. We can just return an anonymous class Object like /* Object obj = new Object(){ //reduced string and number of operations variables}; */

Comment (neerajkujha) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: Can you please explain this anonymous class object.
Not able to comprehend your explanation.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 9 above)
Comment (neerajkujha) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: Can you please explain difference between both with advantages and disadvantages. Cant we use int in map directly instead of Integer?

File: Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java (hunk 14 above)
Comment (neerajkujha) [reply Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java:115]: It is always better to handle null or empty cases in code than throwing exceptions. As name suggests exceptions should be used for exceptional situations.  Handling and validating input should be developer responsibility and handled more elegantly and efficiently

File: solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java (hunk 15 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: anonymous_obj
Hi Neeraj,

Please find the attached image for the object that we can return instead of creating a ChopCount class.

Regards,
Arjun

Comment (neerajkujha) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: Thanks

Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:54]: One more way is to just have those variables in NonReducible class itself and return the same instead of having a separate ChopCount class.

File: solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java (hunk 9 above)
Comment (arjun9218) [reply solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:29]: Integer is a wrapper class right, so it has many additional methods. Primitive datatypes are just basic. Advantage of a primitive datatype is that performance wise it is faster compared to Wrapper classes, and wrapper classes takes up more space compared to primitive datatypes.
We can't use int directly in the map because it accepts only classes derived from Object, int, float etc., primitives are not derived from Object.

File: Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java (hunk 14 above)
Comment (arjun9218) [reply Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java:115]: Hmmm yeah I agree with that. I'll keep that in mind.
